packages/completer/src/reconciliator.ts

```diff
@@ -0,0 +1,241 @@
+  /**
+   * Creates an instance of ProviderReconciliator.
+   */
+  constructor(options: ProviderReconciliator.IOptions) {
+    this._providers = options.providers;
+    this._context = options.context;
+    this._timeout = options.timeout;
+  }
+
+  /**
+   * Fetch response from multiple providers, If a provider can not return
+   * the response for a completer request before timeout,
+   * the result of this provider will be ignored.
+   *
+   * @param {CompletionHandler.IRequest} request - The completion request.
+   */
+  public async fetch(
+    request: CompletionHandler.IRequest
+  ): Promise<CompletionHandler.ICompletionItemsReply | null> {
+    const current = ++this._fetching;
+    let promises: Promise<CompletionHandler.ICompletionItemsReply | null>[] =
+      [];
+    for (const provider of this._providers) {
+      let promise: Promise<CompletionHandler.ICompletionItemsReply | null>;
+      promise = provider.fetch(request, this._context).then(reply => {
+        if (current !== this._fetching) {
+          return Promise.reject(void 0);
+        }
+        const items = reply.items.map(el => ({
+          ...el,
+          resolve: this._resolveFactory(provider, el)
+        }));
+        return { ...reply, items };
+      });
+
+      const timeoutPromise =
+        new Promise<CompletionHandler.ICompletionItemsReply | null>(resolve => {
+          return setTimeout(() => resolve(null), this._timeout);
+        });
+      promise = Promise.race([promise, timeoutPromise]);
+      // Wrap promise and return error in case of failure.
+      promises.push(promise.catch(p => p));
+    }
+    // TODO: maybe use `Promise.allSettled` once library is at es2020 instead of adding a catch.
+    const combinedPromise = Promise.all(promises);
+    return this._mergeCompletions(combinedPromise);
+  }
+
+  private _alignPrefixes(
+    replies: CompletionHandler.ICompletionItemsReply[],
+    minStart: number,
+    maxStart: number
+  ): CompletionHandler.ICompletionItemsReply[] {
+    if (minStart != maxStart) {
+      const editor = this._context.editor;
+      if (!editor) {
+        return replies;
+      }
+      const cursor = editor.getCursorPosition();
+      const line = editor.getLine(cursor.line);
+      if (!line) {
+        return replies;
+      }
+
+      return replies.map(reply => {
+        // No prefix to strip, return as-is.
+        if (reply.start == maxStart) {
+          return reply;
+        }
+        let prefix = line.substring(reply.start, maxStart);
+        return {
+          ...reply,
+          items: reply.items.map(item => {
+            let insertText = item.insertText || item.label;
+            item.insertText = insertText.startsWith(prefix)
+              ? insertText.slice(prefix.length)
+              : insertText;
+            return item;
+          })
+        };
+      });
+    }
+    return replies;
+  }
+
+  private async _mergeCompletions(
```

Comment:
Could you move these two private methods down behind the public methods?

Reply:
Done in 80286edd99eb15ea656802f4d7c87eeaf1fcec77.